### ISSUE TITLE
Linuxdeploy from August 31 does not pack some libraries it used to

### ISSUE THREAD

**Issue by bvschaik** [linuxdeploy/linuxdeploy#145]
For my project [Julius](https://github.com/bvschaik/julius), we package an AppImage, which is created on Travis.

Starting on August 31, the resulting AppImage files are 0.7 MB smaller than they are used to, and they no longer run on, for example, Manjaro, which gives the following error:

    ./julius.AppImage: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory

The AppImage is created on Travis using the following script: https://github.com/bvschaik/julius/blob/master/.ci_scripts/package_appimage.sh

The script, combined with the previous build steps, comes down to:

    mkdir build
    cd build
    cmake ..
    make
    make DESTDIR=AppDir install
    wget https://github.com/linuxdeploy/linuxdeploy/releases/download/continuous/linuxdeploy-x86_64.AppImage
    chmod +x linuxdeploy*.AppImage
    ./linuxdeploy-x86_64.AppImage --appdir AppDir --output appimage

Logs from the old version which results in a working AppImage (linuxdeploy travis build 478): https://travis-ci.com/github/bvschaik/julius/jobs/385719358#L2196
Logs from the newest version which results in an AppImage with missing deps (linuxdeploy travis build 479): https://travis-ci.com/github/bvschaik/julius/jobs/381876754#L2183

The only difference between the two builds is the linuxdeploy version.

Comparing the logs, I see the following libraries being deployed in version 478, but not in 479:

```
Deploying shared library /usr/lib/x86_64-linux-gnu/libvorbisenc.so.2
Deploying shared library /lib/x86_64-linux-gnu/libtinfo.so.5
Deploying shared library /lib/x86_64-linux-gnu/libselinux.so.1
Deploying shared library /lib/x86_64-linux-gnu/liblzma.so.5
Deploying shared library /lib/x86_64-linux-gnu/libgcrypt.so.20
Deploying shared library /lib/x86_64-linux-gnu/libnsl.so.1
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libresolv.so.2 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libgpg-error.so.0
```

Build environment: Ubuntu Xenial LTS (from Travis), also reproducible on a local Xenial virtual machine.
Build dependencies: SDL2, SDL2_mixer

Any clue what's causing this, and how to fix it?

**Comment by TheAssassin** [comment]
The only reason I could come up with is that linuxdeploy's new subprocessing might lose a few lines from the `ldd` output... but to prove that theory, I would need debug logs. Please re-run with `-v0` and post both logs. (*please use the code syntax* or maybe a pastebin).

**Comment by bvschaik** [comment]
- [Log for build 478](https://pastebin.com/KmaBin1t)
- [Log for build 479](https://pastebin.com/ENqrQDfQ)

Interestingly enough, if I build with 479 again without cleaning the App dir, I get a correct AppImage with all libs included.
- [Log for build 479, second run](https://pastebin.com/y1xUDRrD)

The problem might be in these lines, which look to me like a cut-off `ldd` output:
(build 479, first run):
```
DEBUG: Invalid ldd output:  	libvorbisenc.so.2 => /usr/lib/x86_64-linux-gnu/libvor 
```
(build 479, second run):
```
DEBUG: Invalid ldd output:  	libvorbis.so.0 => /home/bianca/julius/build-479/App/usr/bin/../lib/libvorbis.s 
```

**Comment by bvschaik** [comment]
I think commit 1072c6d9f0d0dbed006125077ecd2e272d0c2476 is the issue. Having a buffer of 4096 characters and letting the subprocess run unchecked for 50ms is going to fill up the buffer and discard the remainder. If I run `ldd` on the original executable and take the first 4096 bytes, I end up with these last few lines:
```
...
        libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f14786f1000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f14784db000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f147826b000)
        libvorbisenc.so.2 => /usr/lib/x86_64-linux-gnu/libvor
```

If I run `ldd` on the executable after running linuxdeploy (after it has updated the library paths) and again take 4096 bytes:
```
...
        libpulse-simple.so.0 => /home/bianca/julius/build-479/App/usr/bin/../lib/libpulse-simple.so.0 (0x00007f9e7cdbc000)
        libsndfile.so.1 => /home/bianca/julius/build-479/App/usr/bin/../lib/libsndfile.so.1 (0x00007f9e7cb51000)
        libreadline.so.6 => /home/bianca/julius/build-479/App/usr/bin/../lib/libreadline.so.6 (0x00007f9e7c907000)
        libvorbis.so.0 => /home/bianca/julius/build-479/App/usr/bin/../lib/libvorbis.s
```
This coincides with the "invalid" output lines I posted above, so the ldd output is cut off and therefore the last libraries are not included in the AppImage. For the first run, the omitted libraries are exactly the ones that are cut off from the ldd output. I'm still not sure why it does succeed on the second run though, since the output is larger...

**Comment by TheAssassin** [comment]
So I guess after the process is dead, the contents remaining in the pipe aren't read completely?

**Comment by bvschaik** [comment]
It looks like it, yes: the `proc.is_running()` check exits the for-loop without emptying the pipe.

**Comment by TheAssassin** [comment]
Looks like I got a working patch. First, I reduced the buffer size to 32 bytes. I could reproduce the incomplete ldd output very easily. I then introduced a nested loop there to just repeatedly read until there's no more data left in the pipe. This ensures all available data is read and copied into the large buffer immediately. I tested everything again with the tiny buffer, letting the program print how often the for loop was run. With the tiny buffer size, on some binaries, this loop ran over 100 iterations, and the output was always complete (checked against a manual `ldd` run).

I hope this patch fixes the issue permanently. I'm preparing a PR right now. Looking forward to seeing you test it.
